virttest/cpu.py

```diff
@@ -1258,3 +1261,110 @@ def get_qemu_best_cpu_model(params):
     return params.get("default_cpu_model", None)
+
+
+def get_qemu_best_cpu_info(params):
+    """
+    Get the best cpu model and features available for qemu
+    """
+    cpu_info = {}
+    cpu_model = get_qemu_best_cpu_model(params)
+    # Expand if needed
+    cpu_flags = ""
+    cpu_info["model"] = cpu_model
+    cpu_info["flags"] = cpu_flags
+    return cpu_info
+
+
+def convert_flags_from_str_to_dict(cpu_flags_str):
+    """
+    Convert cpu flags string to a dict
+    """
+    flags_dict = {}
+    if cpu_flags_str:
+        flags_list = cpu_flags_str.lstrip(",").split(",")
+        for item in flags_list:
+            if item.startswith("+"):
+                flags_dict[item[1:]] = "on"
+            elif item.startswith("-"):
+                flags_dict[item[1:]] = "off"
+            elif item.find("=") != -1:
+                item = item.split("=")
+                flags_dict[item[0]] = item[1]
+            else:
+                flags_dict[item] = None
+    return flags_dict
+
+
+def get_cpu_info_from_qemu_cli(name):
+    """
+    Get cpu info from qemu cmdline
+    """
+    cmd = "ps aux | grep qemu | grep -e '\s-name\s\+guest=%s\(\,\|\s\)'" % name
+    output = process.getoutput(cmd, shell=True)
+    cpu_line = re.findall("-cpu\s+(\S*)\s+", output)[0]
+    cpu_model, cpu_flags = cpu_line.split(',', 1)
+    cpu_info = {}
+    cpu_info["model"] = cpu_model
+    cpu_info["flags"] = cpu_flags
+    return cpu_info
+
+
+def get_cpu_info_from_virsh(params):
+    """
+    Try to get cpu model and features from virsh with 'host-model'
+    """
+    vm_arch = params["vm_arch_name"]
+    if vm_arch not in ['x86_64', 'i686']:
+        raise NotImplementedError("Arch '%s' is not supported" % vm_arch)
+    machine = params.get('machine_type')
+    name = uuid.uuid4().hex
+    try:
+        path.find_command("virsh")
+    except path.CmdNotFoundError:
+        logging.warning("Virsh executable not set or found on path")
+        return
+
+    xml = """
+            <domain type='kvm'>
+                <name>%s</name>
+                <memory>1</memory>
+                <os>
+                    <type arch='%s' machine='%s'>hvm</type>
+                </os>
+                <cpu mode='host-model'/>
+            </domain>
+          """ % (name, vm_arch, machine)
+    xml_file = os.path.join(data_dir.get_tmp_dir(), "temp_xml_for_cpu")
+    with open(xml_file, "w") as f:
+        f.write(xml)
+    try:
+        logging.info("Get cpu model and features from virsh")
+        virsh.define(xml_file)
+        virsh.start(name)
+    except Exception as err:
+        logging.error(err)
+        return
+    else:
+        cpu_info = get_cpu_info_from_qemu_cli(name)
+        return cpu_info
+    finally:
+        if virsh.is_alive(name):
+            virsh.destroy(name, ignore_status=True)
+        virsh.undefine(name, ignore_status=True)
+
+
+def combine_flags(auto_cpu_flags, custom_cpu_flags):
```

Comment:
```suggestion
def recombine_qemu_cpu_flags(base, suggestion):
```

Reply:
@luckyh  How to explain the meaning of `base` and `suggestion` ?   On point i can think of is: the priority of `suggestion` is higher than `base`.

Reply:
Yes, it is, as we used `dict.update` at here.